### ISSUE TITLE
Kinematicbody snap drags objects down even without any velocity

### ISSUE THREAD

**Issue by moiman100** [godotengine/godot#22313]
Godot version 534b7ef

**This is 3D, it is just sideview to better illustrate what is happening.**
![snap_drag](https://user-images.githubusercontent.com/40793342/45876405-433a1900-bda3-11e8-92a7-119e0840f08f.gif)

This is the full code in kinematicbody.

```
extends KinematicBody

var vel = Vector3()

func _physics_process(delta):
	vel = move_and_slide_with_snap(vel, Vector3(0, -1, 0), Vector3(0, 1, 0))
```
[Minimal Project.zip](https://github.com/godotengine/godot/files/2404944/Minimal.Project.zip)

**Comment by piratesephiroth** [comment]
You think this is bug?
`stop_on_slope `is set to false by default

**Comment by moiman100** [comment]
Yes but there is no velocity applied to the kinematicbody so it shouldn't move.

**Comment by piratesephiroth** [comment]
it will always slide down any slope if stop_on_slope is set to false.

**Comment by moiman100** [comment]
No, kinematicbody doesn't move anywhere if you don't move it.

**Comment by moiman100** [comment]
Increasing collision safe margin makes it go faster so I think something like this happens:
Kinematicbody snaps straight down to floor, then separates to safe margin along collision normal.
Since collision normal is not parallel to y axis it moves slightly sideways and then this repeats.

**Comment by moiman100** [comment]
> it will always slide down any slope if stop_on_slopeis set to false.
> 
> Remember you're dealing with 3.1 and it has gone through a bunch of undocumented changes. You can't expect it to work exactly like 3.0.

Are you seriously suggesting that it would just magically move?
The only reason it currently moves in the provided minimal project is because the ball starts slightly inside the surface. This collision activates the snapping. If you try moving the ball out of the surface and then running the project the ball will stay still in the air, because there is nothing in the script that is moving it.

**Comment by piratesephiroth** [comment]
And I just noticed that stop_on_slope is broken on 3D too and slows down the sliding speed when moving uphill.

**Comment by piratesephiroth** [comment]
Ah, it's the snapping force pulling you downhill. Try seeting it to a positive value and you'll see that you don't slide at all.

**Comment by capnm** [comment]
> Ah, it's the snapping force pulling you downhill.

If you read the current docs, the vector is just a simple vector (no magical powers involved):

» As long as the **snap vector is in contact with the ground**, the body will remain attached to the surface. This means you must disable snap in order to jump, for example. You can do this by settingsnap to(0, 0, 0) or by using move_and_slide() instead. «

**Comment by piratesephiroth** [comment]
> 
> 
> > Ah, it's the snapping force pulling you downhill.
> 
> If you read the current docs, the vector is just a simple vector (no magical powers involved):
> 
> » As long as the **snap vector is in contact with the ground**, the body will remain attached to the surface. This means you must disable snap in order to jump, for example. You can do this by settingsnap to(0, 0, 0) or by using move_and_slide() instead. «

Well then the vector is pulling the kinematic body straight down on the Y axis (-1) so I suppose it's working like gravity. The surface is sloped too so the body will slide down.

**Comment by capnm** [comment]
> Well then the vector is pulling the kinematic body straight down on the Y axis (-1) so I suppose it's working like gravity. The surface is sloped too so the body will slide down.

If the function doc. is correct, the vector doesn't represent any force. It just selects the slope where the body should be attached. How it's implemented is another question.

**Comment by jedStevens** [comment]
Not in a place to share my project but, 

Changing from 3.0 to 3.1 has affected my player movement and he will now slide down slopes. This is a slight improvement for me but still not perfect.

 I will agree that `stop_on_slope` is messed up, with and without gravity sliding on the slope still occurs. Not sure how to fix this properly but we should stamp the bug label on this and get a proper fix in. This is a core function imo.

**Comment by AndreaCatania** [comment]
I would like to revert stop_on_slope from that function that seems not so good. You can always use the system with ray shape that reduz created https://godotengine.org/article/godot-31-will-get-many-improvements-kinematicbody

What do you think about it?

**Comment by piratesephiroth** [comment]
> 
> 
> I would like to revert stop_on_slope from that function that seems not so good. You can always use the system with ray shape that reduz created https://godotengine.org/article/godot-31-will-get-many-improvements-kinematicbody
> 
> What do you think about it?

I love the raycast shape idea but it was already broken (more like unfinished though) even before stop_on_slope.
A single ray would work perrfectly and in any direction (there was no cast_to property but we could rotate it) but it more than one ray would cause weird errors.

With stop_on_slope enabled the ray shapes behave very diffrently. Multiple rays don't cause errors anymore and they only work when pointing down but they cause some jittery collisiions with the floor.

**Comment by moiman100** [comment]
Well stop_on_slope isn't fully working for me, but this issue is not about it.

**Comment by piratesephiroth** [comment]
(I would vote for reverting back to slope_min_velocity and fixing the raycast shapes)

**Comment by reduz** [comment]
I think this is more of a bug of stop_on_slope, @AndreaCatania can you give a check to this?
Raycast shapes now work, though.

**Comment by piratesephiroth** [comment]
>>Raycast shapes now work, though

They kinda work, unless issue #25050 been addressed already.

**Comment by reduz** [comment]
The solution to this for now probably would be to not use snap if your character is not actually moving left or right (or use a threshold). I'm not quite sure this is a bug yet, but I plan to rewrite both KinematicBodies (2D and 3D) for 4.0 into a way where these things can be customized more. I will kick to then.

**Comment by Scemenzo** [comment]
I've just encountered the exact same problem using the KinematicBody2D node. I was going mad.
I have the exact same behaviour as the one that @moiman100 described.
I've reviewed my code carefully and I'm (now) certain it is this very same Godot's bug.
If it can help: increasing the gravity force (increment on Y coordinate before using the method move_and_slide_with_snap) produce the same sliding behaviour.
It's certainly something that happend when calling the method. Maybe the Vector2/Vector3 returned and put into the valocity variable.

**Comment by Scemenzo** [comment]
Just to add more infos, even if delayed: 
on move_and_slide_with_snap, if stop_on_slope is false the velocity.x value is not zero after the method.
If it is true, velocity.x is zero and the character slides a bit slower.
![immagine](https://user-images.githubusercontent.com/12174888/53729541-cc3a9d80-3e75-11e9-8cb7-055aeb118e8b.png)

**Comment by behelit2** [comment]
Like @Scemenzo , I've also encountered the issue with KinematicBody2D. Kind of annoying, I thought the whole point of snap was to make working with slopes easier. Even with very gentle slopes and stop_on_slope enabled, the kinematic character will continue to move slowly down the hill. Going to try @reduz 's suggestion to use move_and_slide when not walking. Maybe try zeroing out the motion if it falls below a certain number.

**Comment by jitspoe** [comment]
This is kind of a fundamental problem with margins.  Move an object down until it collides, then the margin will push it out along the normal.  It's also (physics) framerate dependent.  The more frames you have, the more frequently it'll get pushed out.  Effectively the same issue as this:

https://github.com/godotengine/godot/issues/18624

**Comment by Bropocalypse** [comment]
As @jitspoe suggests, yes, the problem is that they're pushed out in the direction of the surface normal. What would be ideal, I think, would be to optionally specify a margin based on angle of attack- In other words, separate the objects in the direction of velocity rather than direction of the normal, with that velocity-based separation's length calculated to the point where the objects are separated by a distance equal to the normal-based margin.
